### PR TITLE
OpenBackhaul/issue132

### LINKED ISSUE #132: Off-load topology information into ElasticSearch
The following aspects have to be regarded while switching ALT from storing topology information into its own CONFIGfile towards off-loading topology information into a separate database:

Create ElasticSearchClient
- [x] ServiceList
- [x] CONFIGfile

Adapt modeling structure towards standard
- [x] CONFIGfile
  - [x] remove network-control-domain
  - [x] remove other application's control-construct
  - [x] remove links
- [x] OAS
  - [x] either shorten path statements or add _appData://_ or _ElasticSearch://_ in descriptions of individual services and common components

Stringently define self-calling (assure that ALT as an arbitrary application of the application layer is calling ALT as the storage of the current network topology, in case there are changes at its own interface or forwarding configurations)
- [x] ServiceList
- [x] ForwardingList
- [x] CONFIGfile
- [x] OAS

### PR DESCRIPTION
**Pull request by openBackhaul**
Fixes #132

### CODE REVIEW

**Review by IswaryaaS — CHANGES_REQUESTED** [review]

File: ApplicationLayerTopology.yaml [hunk 1]
```diff
@@ -138,7 +138,7 @@ paths:
...
+          url: #[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-im-nr-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-protocol]://[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-im-nr-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-address/domain-name or /core-model-1-4:network-control-domain/control-construct=alt-2-0-0/logical-termination-point=alt-2-0-0-tcp-c-im-nr-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-address/ip-address/ipv-4-address]:[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-im-nr-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-port][/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-op-c-im-nr-2-0-0-000/layer-protocol=0/operation-client-interface-1-0:operation-client-interface-pac/operation-client-interface-configuration/operation-name]
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml:141]: Changed as proposed.
Many thanks for indicating.

File: ApplicationLayerTopology.yaml [hunk 2]
```diff
@@ -227,249 +227,8 @@ paths:
...
+          url: #[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-bm-ro-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-protocol]://[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-bm-ro-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-address/domain-name or /core-model-1-4:network-control-domain/control-construct=alt-2-0-0/logical-termination-point=alt-2-0-0-tcp-c-bm-ro-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-address/ip-address/ipv-4-address]:[/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-tcp-c-bm-ro-2-0-0-000/layer-protocol=0/tcp-client-interface-1-0:tcp-client-interface-pac/tcp-client-interface-configuration/remote-port][/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-op-c-im-ro-2-0-0-000/layer-protocol=0/operation-client-interface-1-0:operation-client-interface-pac/operation-client-interface-configuration/operation-name]
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml:231]: It's a result of Ctrl+h and it affects 7 occurrences. 
Changed as proposed in all cases.
Thanks for indicating.

File: ApplicationLayerTopology.yaml [hunk 3]
```diff
@@ -1195,52 +954,52 @@ paths:
...
+                            from [/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-http-s-bm-000/layer-protocol=0/http-server-interface-1-0:http-server-interface-pac/http-server-interface-capability/release-number]'
                         topology-operation-application-update:
                           type: string
                           description: >
                             'Operation for updating the entire information about an application
-                            from [/core-model-1-4:network-control-domain/control-construct=alt-2-0-0/logical-termination-point=alt-2-0-0-op-s-is-004/layer-protocol=0/operation-server-interface-1-0:operation-server-interface-pac/operation-server-interface-capability/operation-name]'
+                            from [/core-model-1-4:control-construct/logical-termination-point=alt-2-0-0-op-s-is-004/layer-protocol=0/operation-server-interface-1-0:operation-server-interface-pac/operation-server-interface-capability/operation-name]'
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml:967]: ApplicationPattern has been changed as proposed.
ALT will be updated while executing ALT issue #100 (_After updating all TAC applications: Update Callbacks_).

File: ApplicationLayerTopology.yaml [hunk 4]
```diff
@@ -1770,9 +1525,9 @@ paths:
                   pattern: '^([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-(op|http|tcp)-c-([bi][ms])-([0-9]{3})$'
                   description: >
                     'LtpUuid identifying the LTP of an OperationClient, HttpClient or TcpClient, which is to be deleted
-                    find in [/core-model-1-4:network-control-domain/control-construct=*/logical-termination-point=*/uuid]'
+                    find in [ElasticSearch://core-model-1-4:network-control-domain/control-construct=*/logical-termination-point=*/uuid]'
               example:
-                uuid: 'alt-2-0-0-op-s-bm-002'
+                uuid: 'cc-1-0-0-op-s-bm-002'
```
Comment (PrathibaJee) [inline ApplicationLayerTopology.yaml:1530]: As per the description , this service "Deletes an OperationClient, HttpClient or TcpClient LTP and all its dependents".
Proposing to change the example from 
cc-1-0-0-op-s-bm-002
to 
cc-1-0-0-op-c-bm-002

**Review by PrathibaJee — CHANGES_REQUESTED** [review]
Provided few comments which are not related to the current pull request. Please consider them if that is not planned in the roadmap.
Apart from that , a generic doubt , 
The new uuid structure will be effective from version 2 and no application in version 1 will have the format "*-1-0-0-http-s-*-000"
Can we not make the version static here. Kindly let me know.

**Review by IswaryaaS — DISMISSED** [review]
The changes looks fine for removing network-domain from ALT. Thus approving the pull_request

File: ApplicationLayerTopology+services.yaml [hunk 5]
```diff
@@ -82,13 +82,13 @@ servers:
   http-server:
     own-application-name: ApplicationLayerTopology
     own-release-number: 2.0.0
-    uuid: alt-2-0-0-http-s-is-000
+    uuid: alt-2-0-0-http-s-bm-000
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology+services.yaml:85]: ApiSegment shall be removed from HttpClients/Servers and TcpClients/Servers.
Separate issues will be create for updating the guidelines and specifications of ApplicationPattern and all applications of the TAC.

File: ApplicationLayerTopology+forwardings.yaml [hunk 6]
```diff
@@ -304,25 +274,6 @@ forwardings:
...
-          uuid: alt-2-0-0-op-s-bm-009
-      fc-port-deletion:
-      operation-client-deletion:
-    initiating-requests:
-      - server-name: /v1/redirect-topology-change-information
-        uuid: alt-2-0-0-op-s-bm-009
-    consequent-requests:
-      - client-name: ApplicationLayerTopology://v1/update-all-ltps-and-fcs
-        uuid: alt-2-0-0-op-c-bm-alt-2-0-0-000
-
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology+forwardings.yaml:325]: Agree with your point.
As you write above, this aspect is not directly related with off-loading topology information into ElasticSearch.
I created a separate ALT issue #124 and will create further issues in the other applications' repositories.

File: ApplicationLayerTopology.yaml [hunk 7]
```diff
@@ -2072,7 +1827,7 @@ paths:
                   pattern: '^([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-op-fc-([bi][ms])-([0-9]{3})$'
                   description: >
                     'FcUuid identifying the FC, which is to be updated
-                    find or create [/core-model-1-4:network-control-domain/control-construct=*/forwarding-domain=*-1-0-0-op-fd-000/forwarding-construct=*-1-0-0-op-fc-*/uuid]'
+                    find or create [ElasticSearch://core-model-1-4:network-control-domain/control-construct=*/forwarding-domain=*-1-0-0-op-fd-000/forwarding-construct=*-1-0-0-op-fc-*/uuid]'
```
[the remark this answers is not in the document]
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml:1830]: Changed as proposed.

File: ApplicationLayerTopology.yaml (hunk 4 above)
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml:1530]: Changed to _cc-1-0-0-op-c-bm-alt-2-0-0-000_.
Thanks for indicating.

File: ApplicationLayerTopology.yaml [hunk 8]
```diff
@@ -2836,25 +2591,25 @@ paths:
                           type: string
                           description: >
                             'Name of application that holds the link end point (operationServer or operationClient)
-                            from [/core-model-1-4:network-control-domain/control-construct=*/logical-termination-point=alt-2-0-0-http-*-*/layer-protocol=0/http-*-interface-1-0:http-*-interface-pac/http-*-interface-*/application-name]'
+                            from [ElasticSearch://core-model-1-4:network-control-domain/control-construct=*/logical-termination-point=alt-2-0-0-http-*-*/layer-protocol=0/http-*-interface-1-0:http-*-interface-pac/http-*-interface-*/application-name]'
```
[the remark this answers is not in the document]
Comment (PrathibaJee) [reply ApplicationLayerTopology.yaml:2594]: Yes Thorsten. absolutely fine.

File: ApplicationLayerTopology.yaml (hunk 4 above)
Comment (PrathibaJee) [reply ApplicationLayerTopology.yaml]: Sorry for not mentioning the correct uuid format. Many thanks for correcting it.

File: ApplicationLayerTopology.yaml [hunk 9]
```diff
@@ -1770,9 +1513,9 @@ paths:
                   pattern: '^([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-(op|http|tcp)-c-([bi][ms])-([0-9]{3})$'
```
Comment (PrathibaJee) [inline ApplicationLayerTopology.yaml:1513]: Sorry for missing it before. The pattern needs to be corrected.
from
^([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-(op|http|tcp)-c-([bi][ms])-([0-9]{3})$
to 
^([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-(op|http|tcp)-c-([bi][ms])-([a-z]{2,6})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{1,2})-([0-9]{3})$

**Review by PrathibaJee — CHANGES_REQUESTED** [review]
Apart from the mentioned 1 change , no further comments.
Many thanks.

File: ApplicationLayerTopology.yaml (hunk 9 above)
Comment (openBackhaul) [reply ApplicationLayerTopology.yaml]: Thanks for indicating.
Pattern had to be adapted to the latest decision (removing ApiSegment from http and tcp client and server UUIDs), too.

**Review by PrathibaJee — APPROVED** [review]
Thanks for fixing the LTP uuid format. Its working fine now. 
No further comments. So approving the pull request.

**Pull request merged**